orchagent/pfcwdorch.cpp

```diff
@@ -0,0 +1,755 @@
+        {
+            nearestTime = queueEntry.pollTimeLeft;
+        }
+    }
+
+    return nearestTime;
+}
+
+template <typename DropHandler, typename ForwardHandler>
+void PfcWdSwOrch<DropHandler, ForwardHandler>::pollQueues(uint32_t nearestTime, DBConnector& db,
+        string detectSha, string restoreSha)
+{
+    SWSS_LOG_ENTER();
+
+    unique_lock<mutex> lk(m_pfcWdMutex);
+
+    // Select those queues for which timer expired
+    vector<string> normalQueues;
+    vector<string> stormedQueues;
+    for (const auto& queueKv : m_entryMap)
+    {
+        sai_object_id_t queueId = queueKv.first;
+        const PfcWdQueueEntry& queueEntry = queueKv.second;
+
+        if (queueEntry.pollTimeLeft == nearestTime)
+        {
+            // Queue is being stormed
+            if (queueEntry.handler != nullptr)
+            {
+                stormedQueues.push_back(sai_serialize_object_id(queueId));
+            }
+            // Queue is not stormed
+            else
+            {
+                normalQueues.push_back(sai_serialize_object_id(queueId));
+            }
+        }
+    }
+
+    // Run scripts for selected queues to see if their state changed
+    // from normal to stormed and vice versa
+    set<string> stormCheckReply;
+    set<string> restoreCheckReply;
+    vector<string> argv =
+    {
+        to_string(COUNTERS_DB),
+        COUNTERS_TABLE,
+        to_string(nearestTime * 1000)
+    };
+
+    if (!normalQueues.empty())
+    {
+        stormCheckReply = runRedisScript(db, detectSha, normalQueues, argv);
+    }
+
+    if (!stormedQueues.empty())
+    {
+        restoreCheckReply = runRedisScript(db, restoreSha, stormedQueues, argv);
+    }
+
+    // Update internal state of queues and their time
+    for (auto& queueKv : m_entryMap)
+    {
+        sai_object_id_t queueId = queueKv.first;
+        PfcWdQueueEntry& queueEntry = queueKv.second;
+
+        string queueIdStr = sai_serialize_object_id(queueId);
+
+        // Queue became stormed
+        if (stormCheckReply.find(queueIdStr) != stormCheckReply.end())
+        {
+            if (queueEntry.c_action == PfcWdAction::PFC_WD_ACTION_DROP)
+            {
+                queueEntry.handler = make_shared<DropHandler>(
+                        queueEntry.portId,
+                        queueId,
+                        queueEntry.index,
+                        PfcWdOrch<DropHandler, ForwardHandler>::getCountersTable());
+            }
+            else if (queueEntry.c_action == PfcWdAction::PFC_WD_ACTION_FORWARD)
+            {
+                queueEntry.handler = make_shared<ForwardHandler>(
+                        queueEntry.portId,
+                        queueId,
+                        queueEntry.index,
+                        PfcWdOrch<DropHandler, ForwardHandler>::getCountersTable());
+            }
+            else
+            {
+                throw runtime_error("Invalid PFC WD Action");
```

Comment:
Are we catching this exception?

Reply:
No. Its an assertion. We never have to end up in this branch of execution.

Reply:
I don't think it worth to crash the whole orch for this. Can we log the error instead?

Reply:
It is a common way in orchagent to handle exceptional situations that we shouldn't reach in normal flow.